File: src/javascript/algorithms/strings/betterStringCompression/betterStringCompression.js
function betterCompression(compressed) {
    const freq = {};
    const order = [];
    let i = 0;
    while (i < compressed.length) {
        const ch = compressed[i++];
        let num = 0;
        while (i < compressed.length && /[0-9]/.test(compressed[i])) {
            num = num * 10 + Number(compressed[i]);
            i++;
        }
        if (!(ch in freq)) {
            freq[ch] = num;
            order.push(ch);
        } else {
            freq[ch] += num;
        }
    }
    return order.map(ch => ch + freq[ch]).join('');
}

module.exports = { betterCompression };

File: src/python/algorithms/strings/betterStringCompression/better_string_compression.py
def better_compression(compressed):
    freq = {}
    order = []
    i = 0
    n = len(compressed)
    while i < n:
        ch = compressed[i]
        i += 1
        num = 0
        while i < n and compressed[i].isdigit():
            num = num * 10 + int(compressed[i])
            i += 1
        if ch not in freq:
            freq[ch] = num
            order.append(ch)
        else:
            freq[ch] += num
    return ''.join(f"{ch}{freq[ch]}" for ch in order)
